riscv/processor.cc

```diff
@@ -609,18 +630,23 @@ void processor_t::set_csr(int which, reg_t val)
   reg_t delegable_ints = supervisor_ints | coprocessor_ints;
   reg_t all_ints = delegable_ints | MIP_MSIP | MIP_MTIP;
 
-  if (which >= CSR_PMPADDR0 && which < CSR_PMPADDR0 + state.n_pmp) {
-    size_t i = which - CSR_PMPADDR0;
+  size_t i = which - CSR_PMPADDR0;
+  if (which >= CSR_PMPADDR0 && which < CSR_PMPADDR0 + state.platform_n_pmp) {
     bool locked = state.pmpcfg[i] & PMP_L;
-    bool next_locked = i+1 < state.n_pmp && (state.pmpcfg[i+1] & PMP_L);
-    bool next_tor = i+1 < state.n_pmp && (state.pmpcfg[i+1] & PMP_A) == PMP_TOR;
+    bool next_locked = i+1 < (size_t)state.platform_n_pmp && (state.pmpcfg[i+1] & PMP_L);
+    bool next_tor = i+1 < (size_t)state.platform_n_pmp && (state.pmpcfg[i+1] & PMP_A) == PMP_TOR;
     if (!locked && !(next_locked && next_tor))
-      state.pmpaddr[i] = val & ((reg_t(1) << (MAX_PADDR_BITS - PMP_SHIFT)) - 1);
+      state.pmpaddr[i] = val & ((reg_t(1) << (MAX_PADDR_BITS - state.pmp_granularity)) - 1);
 
     mmu->flush_tlb();
+  } else if(which >= CSR_PMPADDR0 + state.platform_n_pmp &&  which < CSR_PMPADDR0 + state.n_pmp) {
+    fprintf(stderr, "error: trying to set PMP (%ld), but the system has %d PMP regions \n",
+            i, state.platform_n_pmp);
+    abort();
```

Comment:
Following the convention of the rest of this function, you should just delete the `else if` part and let this case fall through.